### ISSUE TITLE
Consider simplifying how dense matrix is accessed in HiOp interface 

### ISSUE THREAD

**Issue by pelesh** [LLNL/hiop#81]
HiOp is using [array of row pointers](https://github.com/LLNL/hiop/tree/master/src/Interface) to access dense matrix data through its interface (e.g. constraint Jacobian data). Perhaps more flexible solution, especially for GPU implementations, would be to pass a raw pointer to matrix data. 

Such change would imply all matrix rows are stored in a contiguous memory block. This assumption is actually made at several places in HiOp code, so passing pointer to data block instead of an array of row pointers would make it more clear to user how to store Jacobian data, in addition to making it more GPU friendly.

**Comment by cnpetra** [comment]
for the `double** M` structure you're referring to, the pointer to the (contiguous) block is `M[0]` 

I kinda like `double**` because allows the user do "natural" double indexing, `M[i[j]`...

**Comment by ashermancinelli** [comment]
I agree the 'natural' indexing is nice, but when operating on device pointers, we often want to do something like:
```cpp
rm.copy(JacD[0], Md_->local_buffer(), ns_*nd_*sizeof(double));
```
where JacD is a ptr to matrix buffer. This won't work however, since we're dereferencing device data on host. These calls will work for now (in UVM), but for using device memory in FY21 I think this will become a nontrivial stumbling block.

**Comment by cnpetra** [comment]
I understand now how dereferencing on host is not portable. We can change to single pointers. Changes are throughout all the code tiers, from interface to linear solver backend, but not complicated at all.

**Comment by cnpetra** [comment]
addressed by #128
